### ISSUE TITLE
Request for inclusion: DESLib

### ISSUE THREAD

**Issue by luizgh** [scikit-learn-contrib/scikit-learn-contrib#36]
Request for project inclusion in scikit-learn-contrib

* Project name: DESLib
* Project description: DESlib is an ensemble learning library focusing the implementation of the state-of-the-art techniques for dynamic classifier selection and dynamic ensemble selection.
* Authors: Rafael M. O. Cruz, Luiz G. Hafemann
* Current repository: https://github.com/Menelau/DESlib/
* Requirements:
 * [x] scikit-learn compatible (check_estimator passed): All classifiers passing check_estimator tests
 * [x] Documentation (guide, API reference, example gallery): https://deslib.readthedocs.io/
 * [x] Unit tests (coverage: 95%) 
 * [x] Python3 compatible
 * [x] PEP8 compliant: Compliant, except for E501 (line too long) -  we use a limit of 120 characters
 * [x] Continuous integration: Travis-CI for code (443 unit/integration tests), CircleCI for documentation

**Comment by GaelVaroquaux** [comment]
Overall, with a very cursory look, this seems great.

Two comments:

* Could you add a bit of plotting to your examples? It makes it much
  easier to understand them.

* The limit of 80 chars per line is part of PEP8 and not respecting it
  means that your code deviates very strongly from the standard. As a 
  consequence it requires a different eye (hence switching across
  codebases is tiring) and copy-pasting it to another codebase is not
  possible. You should abide this rule.

Thanks! This is a nice contribution!

**Comment by luizgh** [comment]
Thanks for the quick review. I understand the reasons for the PEP8 compliance, and we will update the code. We will also include examples that are more visual (we already have material that was developed for a tutorial, that we will merge on this library: https://github.com/Menelau/synthetic_datasets). I will let you know when these changes are done

**Comment by luizgh** [comment]
@GaelVaroquaux, we have made the code fully compliant to PEP8 (```pycodestyle deslib``` returns no errors). The gallery of examples was also re-worked, including plots for 6 out of the 9 examples: https://deslib.readthedocs.io/en/latest/auto_examples/index.html
We would appreciate if you can continue with the review for including this package as part of scikit-learn-contrib. Thanks!

**Comment by agramfort** [comment]
+1 on my side. I did run pytest and I can confirm that check_estimator is called on the estimators.

**Comment by ngoix** [comment]
Looks great!
Would be nice to see more benchmarks. I would also be very interested in seeing benchmarks against a stacked model with two layers (same pool of classifiers + decision tree or logistic regression) which is from my understanding the most basic dynamic classifier selection you can do (is it?).

**Comment by luizgh** [comment]
@ngoix - we have added benchmarks against a stacked model on the examples from the gallery. 

Note that stacking is not considered dynamic selection in the DS literature: in stacking the meta-classifier (combiner) receives the output of all base classifiers and aggregates all of them to give the final decision, while in DS the "incompetent" base classifiers are removed first and them only the outputs of the "competent" ones (for the given query sample) is passed to the combiner to get the final answer. Nevertheless, we found that this comparison would be interesting for the community, so we added it to a few examples:

https://deslib.readthedocs.io/en/latest/auto_examples/plot_comparing_dynamic_static.html
https://deslib.readthedocs.io/en/latest/auto_examples/plot_random_forest.html
https://deslib.readthedocs.io/en/latest/auto_examples/plot_example_P2.html

We also fixed the data-leaking issue you found in one of the examples (thanks for spotting that)

Please let us know if there is anything else missing for the inclusion of this project in sklearn-contrib. Thanks!

**Comment by luizgh** [comment]
Hi all, is this request still being reviewed? It has been 20 days since the last update, just want to make sure it doesn't fall into a limbo.

**Comment by ngoix** [comment]
Thanks for the new example @luizgh.

I'm +1 for merging.

Am I right to say that the big difference between 2-layers stacking and DS is that, in DS, the meta-estimator is able to perform "local feature selection" (where 'features'= original features + 1st layer model predictions, and 'local' means nearest neighbors)? 
It would be nice to have an example illustrating the advantage of DS compared to a simple KNeighborsClassifier (I saw the DS vs. KNN example, but it is more about a switching mechanism between DS and KNN, not a proper comparison).

This is a remark for later, I think that the code is good enough to be included as it is.

**Comment by luizgh** [comment]
Thanks for the review @ngoix - can you please advise on the next steps to have this moved to scikit-learn-contrib?

About your question: in DS, the meta-estimator uses the concept of a "Region of competence" that is local (where local can be defined by clustering or nearest neighbors in either input space or in "output profiles") to select or weigh the decisions of the base classifiers. (more details can be found in [1]). We plan on adding more examples to compare DS with other techniques in upcoming releases.

[1] Cruz, R.M., Sabourin, R. and Cavalcanti, G.D., 2018. Dynamic classifier selection: Recent advances and perspectives. Information Fusion, 41, pp.195-216.  https://www.etsmtl.ca/Unites-de-recherche/LIVIA/Recherche-et-innovation/Publications/Publications-2017/RCruz_InfoFusion.pdf

**Comment by GaelVaroquaux** [comment]
+1 for inclusion.

The only way that I know for inclusion is that you add me (or another member of scikit-learn-contrib) to the admins of your repo, and I transfer the repo to scikit-learn-contrib.

**Comment by jnothman** [comment]
If you make them an org member of scikit-learn-contrib you can ask them to
transfer it here.

**Comment by GaelVaroquaux** [comment]
> If you make them an org member of scikit-learn-contrib you can ask them to transfer it here.

Happy to do that: we need more responsible people member of the org.

@luizgh, do volunteer?

**Comment by luizgh** [comment]
Hi @GaelVaroquaux, just to make sure I understand your question - do you mean adding me as a member of scikit-learn-contrib (e.g. for reviewing requests to add projects to the repository. Is there anything else?)

**Comment by GaelVaroquaux** [comment]
> Hi @GaelVaroquaux, just to make sure I understand your question - do you mean
> adding me as a member of scikit-learn-contrib (e.g. for reviewing requests to
> add projects to the repository. Is there anything else?)

All correct.

**Comment by luizgh** [comment]
Sure, I will be glad to do it.

**Comment by luizgh** [comment]
Thanks @GaelVaroquaux - I transferred the repository. @GaelVaroquaux, @ngoix, @jnothman, @agramfort - thanks a lot for taking the time to review this project inclusion!

I should have some time after the holidays - I will check the list of new projects after the new year. Happy holidays!
